Improve diff inline highlighting using per-character/word diff

ychin opened the pull request:
*Problem*:  Diff mode's inline highlighting is lackluster. It only performs a line-by-line comparison, and calculates a single shortest range within a line that could encompass all the changes. In lines with multiple changes, or those that span multiple lines, this approach tends to end up highlighting much more than necessary.

*Solution*: Implement new inline highlighting modes by doing per-character or per-word diff within the diff block, and highlight only the relevant parts.

This change introduces a new diffopt option "inline:<type>". Setting to "none" will disable all inline highlighting, "simple" (the default) will use the old behavior, "char" / "word" will perform a character/word-wise diff of the texts within each diff block and only highlight the differences.

The new char/word inline diff only use the internal xdiff, and will respect diff options such as algorithm choice, icase, and misc iwhite options. indent-heuristics is always on to perform better sliding.

For character highlight, a post-process of the diff results is first applied before we show the highlight. This is because a naive diff will create a result with a lot of small diff chunks and gaps, due to the repetitive nature of individual characters. The post-process is a heuristic-based refinement that attempts to merge adjacent diff blocks if they are separated by a short gap (1-3 characters), and can be further tuned in the future for better results. This process results in more characters than necessary being highlighted but overall less visual noise.

For word highlight, always use first buffer's iskeyword definition. Otherwise if each buffer has different iskeyword settings we would not be able to group words properly.

The char/word diffing is always per-diff block, not per line, meaning that changes that span multiple lines will show up correctly. Added/removed newlines are not shown by default, but if the user has 'list' set (with "eol" listchar defined), the eol character will be be highlighted correctly for the specific newline characters.

Also, add a new "DiffTextAdd" highlight group linked to "DiffText" by default. It allows color schemes to use different colors for texts that have been added within a line versus modified.

This doesn't interact with linematch perfectly currently. The linematch feature splits up diff blocks into multiple smaller blocks for better visual matching, which makes inline highlight less useful especially for multi-line change (e.g. a line is broken into two lines). This could be addressed in the future.

As a side change, this also removes the bounds checking introduced to diff_read() as they were added to mask existing logic bugs that were properly fixed in #16768.

---

# Notes

~~Marking this as draft/WIP as I am still fixing up a couple minor issues, and need to write a fair amount of tests to cover this feature. That said, the feature should be fully usable now (other than `diff_hlID()` not working with this yet), and I would love it if people could give it a try and provide feedbacks.~~ (Update: This PR is no longer draft)

Some extra notes below for those interested:

Some examples of what it looks like:

* `inline:simple` (old behavior)
  * <img width="828" alt="image" src="https://github.com/user-attachments/assets/f6380e67-e0bc-4b49-a449-b780723b521a" />
* `inline:char`
  * <img width="828" alt="image" src="https://github.com/user-attachments/assets/9fb05ccb-4063-4c85-9ead-e70e6f7a96f1" />
* `inline:char,algorithm:patience,icase`
  * <img width="828" alt="image" src="https://github.com/user-attachments/assets/3c6a5f1b-3388-4069-b0bd-5ab7764a71be" />
* `inline:word`
  * <img width="828" alt="image" src="https://github.com/user-attachments/assets/2e40d342-31d9-4780-a9af-4e523e9323f8" />

More complicated multi-line change (the diff comes from this PR itself):

<img width="1108" alt="image" src="https://github.com/user-attachments/assets/63c022b0-d645-4be7-a24a-a749c55d3cdb" />

Multi-buffer highlight:

<img width="463" alt="image" src="https://github.com/user-attachments/assets/c534134a-533e-4df6-8006-036d0937328d" />

Color schemes can define `DiffTextAdd` to highlight added texts differently from changed texts:

<img width="776" alt="image" src="https://github.com/user-attachments/assets/62a23dcc-f801-4a1c-989f-216ff9dc5399" />

## Name

Every diff program seems to call this feature something different. I mostly settled on "inline highlight" but open to other suggestions.

## character diff

As mentioned above, I added a refinement post-process step to clean up the diffs. Just to show the problem and the before/after (first image is a naive implementation, and second is the one with a post-process step to merge small gaps):

<img width="1500" alt="image" src="https://github.com/user-attachments/assets/e34f6e85-5d8d-4834-9f9c-8d3586886435" />
<img width="1500" alt="image" src="https://github.com/user-attachments/assets/d8bc2c50-42d9-4426-85ca-1d8660d65d57" />

From surveying other popular diff programs, the ones that use character diff for inline highlighting (e.g. Meld, VSCode) all have to do similar work, via pre/post-processing to massage the diff results a little bit. VSCode for example does quite a bit of post-processing (merging small gaps, extending small diffs to word boundaries, etc), and I tried to just keep this minimal and go for the largest bang for the buck to make the result look decent without over-complicating the implementation (implemented in `diff_refine_inline_char_highlight()`). This step is heuristics based and could be changed in the future.

### Sliders

Another issue is sliders, where you could slide a diff left or right due to repeating texts. To fix this, I'm always forcing on indent-heuristics to piggyback on it for free white-space sliding alignment. FWIW I think indent-heuristics should be in the default `diffopt` setting value (since Git has had it on by default for a long time) but that's another discussion. Without / with indent-heuristics:

Without indent-heuristics (you can see how they don't align at whitespace boundaries):
<img width="1186" alt="image" src="https://github.com/user-attachments/assets/a7bccfba-4e2a-42dc-98c4-9c48fb305da6" />
With indent-heuristics forced on:
<img width="1186" alt="image" src="https://github.com/user-attachments/assets/a98e0ff7-c6b2-4548-a624-be4fba6625c0" />

Note that this is not perfect. Since we are hijacking a feature designed for line diffing (where xdiff uses the white space and indentation of lines as a heuristics) rather than per-character diff, this doesn't help solve sliders at symbols (ideally the highlight should border the "+" sign):

<img width="663" alt="image" src="https://github.com/user-attachments/assets/4e43b279-6ee3-4562-baef-51b9c37f5c5d" />

If we want to fix this we should probably modify xdiff itself to add a new heuristics mode that's aware that it's doing character diffing and has heuristics scoring designed for intraline diff (e.g. VSCode has a scoring system that rank whitespace, symbols, keywords, etc differently to slide the diff to the best location).

## Word diff

This basically just splits the diff block along word difference instead of character. For the most part it works fine. As mentioned I'm defining a word based on the first buffer's `iskeyword` setting to have a consistent definition. Some other programs that use word diff (e.g. diffchar.vim or Git's own word-diff feature) allows you to define words in different ways or via a regex pattern. For first pass I would rather just keep it simple and just provide a single `inline:word` option for now. We could always add a new `inline:pattern` etc setting later as it's extensible.

## Newline

As noted above, newlines can be diffed as well, *but* it will only show up when `list` option is set (note how the first newline on the right is highlighted):

<img width="628" alt="image" src="https://github.com/user-attachments/assets/05c57928-1b12-4464-9055-80f0d2870e09" />

I was debating back and forth how to show this information. VSCode for example highlights the entire rest of the line in the changed color, which I find to be too much. Meld essentially automatically shows the equivalent of `listchar` only where a change happens. I think this could be an interesting future extension, something along the lines of `set diffopt+=auto-list` where we automatically show `list` only when there is a diff highlight.

There's also an issue here where more than half of the color schemes (including bundled ones) I tried tend to not work well when `list` is on during diffing. I am still not sure whether this is because of how Vim combines the highlight attributes or it's the color schemes' fault. This is usually particularly problematic if the color scheme uses `gui=inverse` for NonText group. That's why most of the screenshots I took were using the third-party "iceberg" color scheme as it works the best in this situation.

## linematch

Unfortunately this PR doesn't work perfectly with linematch as mentioned above. Here's a showcase of the issue:

linematch off:
<img width="1660" alt="image" src="https://github.com/user-attachments/assets/545e68f9-b379-4f21-9e3d-d69a7927847e" />
linematch on:
<img width="1660" alt="image" src="https://github.com/user-attachments/assets/7b82e5bb-9911-4d12-86ab-870ac6a1ef28" />

Note how in the linematch example, the code splits the diff block into two, and therefore inline highlighting fails to recognize that the first part of the texts ("another line of text…" corresponds to the other buffer) and it's marked as changed instead. I have some ideas how to fix it but for scope issues decided not to tackle it for now. (This feature is more complicated when 3-way diff is involved and the "correct" fix is not obvious and may require some additional improvements to multi-buffer diff to work well IMO)

# Anticipated FAQs

## Why not just take Git's word-diff code when we use xdiff already?

Word diff is an application layer feature in Git, not part of the xdiff library. It does something similar to this PR but is specific to Git. It isn't something we can just take.

## Why not a plugin?

There's an existing plugin ([diffchar.vim](https://github.com/rickhowe/diffchar.vim)) that already does something similar. It does interesting things but it's limited by Vim's API. For example, it does not do live update when typing, does not handle multi-line blocks, and does not support multi-buffer diff'ing (this PR handles it fine as it uses Vim's builtin diff block merging in `diff_read`). Vim also has more internal knowledge and can control how the rendering is done instead of a plugin needing to bend over backwards to get this to work.

I also think this feature is a basic requirement for a competent diff program and should be part of the overall design. We should not force users to install a plugin to get access to such features.

It does have some interesting features, such as the ability to highlight corresponding texts in the other buffer when cursor is on a highlighted chunk; and for chunks where there's an added piece of text (meaning there's no corresponding texts in the other buffer), use underline in neighbor texts to highlight where it is inserted in the other buffer. I decided not to clone all of them in this PR.

I do think later on Vim could expose more APIs both for accessing the diff highlight information, and/or provide ways to modify it. E.g. If another diff plugin like Difftastic (which does syntax diffing using treesitter) wants to hook into Vim/Neovim, it should have an API where it could define where exactly the highlight should be. We could add something like `set diffopt+=inline:expr`.

## Any performance issues?

I was planning to do some benchmarks but from initial testing, I simply didn't see any performance issues at all and therefore decided not to do it. Vim will only calculate inline difference  for the on-screen diff blocks, and the result is cached inside the diff block, and therefore in vast majority of cases this change will not cause much performance regression when it's on. It will only be an issue if we have some serious degenerate cases where a single line is really long, in which case from experiences this could be *slightly* slow, but usually still within reason. Xdiff is also quite fast compared to other diff programs so I don't think this should cause much issues.

# Relevant links

https://github.com/neovim/neovim/issues/29549
https://github.com/neovim/neovim/issues/3433
https://github.com/rickhowe/diffchar.vim

---

# TODOs

Current TODOs before PR is fully ready:

- [x] Fix `diff_hlID()` to work properly
- [x] Implement tests

---

There are further diff changes / improvements I would like to work on, but probably will wait to see how this PR goes first. Future potential options include making this work better with linematch, move detection, expose xdiff's `--anchored` feature, better API integration, etc.

ychin committed:
Improve diff inline highlighting using per-character/word diff

Problem:  Diff mode's inline highlighting is lackluster. It only
          performs a line-by-line comparison, and calculates a single
          shortest range within a line that could encompass all the
          changes. In lines with multiple changes, or those that span
          multiple lines, this approach tends to end up highlighting
          much more than necessary.

Solution: Implement new inline highlighting modes by doing per-character
          or per-word diff within the diff block, and highlight only the
          relevant parts.

This change introduces a new diffopt option "inline:<type>". Setting to
"none" will disable all inline highlighting, "simple" (the default) will
use the old behavior, "char" / "word" will perform a character/word-wise
diff of the texts within each diff block and only highlight the
differences.

The new char/word inline diff only use the internal xdiff, and will
respect diff options such as algorithm choice, icase, and misc iwhite
options. indent-heuristics is always on to perform better sliding.

For character highlight, a post-process of the diff results is first
applied before we show the highlight. This is because a naive diff will
create a result with a lot of small diff chunks and gaps, due to the
repetitive nature of individual characters. The post-process is a
heuristic-based refinement that attempts to merge adjacent diff blocks
if they are separated by a short gap (1-3 characters), and can be
further tuned in the future for better results. This process results in
more characters than necessary being highlighted but overall less visual
noise.

For word highlight, always use first buffer's iskeyword definition.
Otherwise if each buffer has different iskeyword settings we would not
be able to group words properly.

The char/word diffing is always per-diff block, not per line, meaning
that changes that span multiple lines will show up correctly.
Added/removed newlines are not shown by default, but if the user has
'list' set (with "eol" listchar defined), the eol character will be be
highlighted correctly for the specific newline characters.

Also, add a new "DiffTextAdd" highlight group linked to "DiffText" by
default. It allows color schemes to use different colors for texts that
have been added within a line versus modified.

This doesn't interact with linematch perfectly currently. The linematch
feature splits up diff blocks into multiple smaller blocks for better
visual matching, which makes inline highlight less useful especially for
multi-line change (e.g. a line is broken into two lines). This could be
addressed in the future.

As a side change, this also removes the bounds checking introduced to
diff_read() as they were added to mask existing logic bugs that were
properly fixed in #16768.

ychin committed:
Fix multi-buffer diff to work

ychin committed:
Fix multi-buffer diff to only use the non-empty blocks

Previously in multi-buffer diff if a buffer doesn't have text in a
particular block, it would cause all the texts in this block to show up
as modified.

ychin committed:
Fix diff_hlID() to work with inline diff

ychin committed:
Fix test_options to pass

ychin committed:
Fix diff_hlID to respect diff flags change

This helps making sure it will return the correct results when
inline:simple changed to inline:none, but also previously it wouldn't
respect changes to add/remove icase due to the caching. In general the
caching probably does more harm than good.

ychin committed:
Fix normal diff icase handling to correctly handle composing chars

This is a drive-by change to fix current bug where a character with
composing character will not have icase work properly since the code was
ignoring it.

ychin committed:
Fix inline diff to not assume first buffer is populated

Vim just fills buffers as they go and don't re-order them if a buffer is
removed.

ychin committed:
Fix icase logic to handle different byte lengths correctly

This happens when we have Å and å. Note that this doesn't fix normal
diffing regarding such texts and simple highlighting. That could be
fixed in the future.

ychin committed:
Add tests for inline highlight

ychin committed:
Fix iskeyword to use the first actual populated non-null buffer, not buffer 0

ychin committed:
Fix memory issues

Properly initialize df_changes to avoid uninitialize memory crash

Also, properly initialize df_lnum / df_count to clear them to zero.
We could make sure to never use the invalid values if the buffer is not
NULL but in practice that is quite tricky and is prone to error making
it easy to do memory bugs. Just set everything to 0 to make it easy to
reason about.

ychin committed:
Clean up ASAN issues

Fix overflow when parsing change

Don't leak memory when diffing multiple buffers. Need to clear the dout
after every diff_read()

ychin committed:
Don't use initializer for struct and use CLEAR_FIELD to make appveyor happy

zeertzjq commented on runtime/doc/options.txt:
```diff
@@ -4392,7 +4407,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 				     W:WildMenu,f:Folded,F:FoldColumn,
 				     A:DiffAdd,C:DiffChange,D:DiffDelete,
-				     T:DiffText,>:SignColumn,-:Conceal,
+				     T:DiffText,E:DiffTextAdd,>:SignColumn,-:Conceal,
```
This line is too long

zeertzjq commented on src/testdir/test_diffmode.vim:
```diff
@@ -2350,6 +2373,162 @@ func Test_diff_topline_noscroll()
 endfunc
 
+" Test inline highlighting which shows what's different within each diff block
+funct Test_diff_inline()
```
```suggestion
func Test_diff_inline()
```

zeertzjq commented on src/testdir/test_diffmode.vim:
```diff
@@ -2350,6 +2373,162 @@ func Test_diff_topline_noscroll()
+  call StopVimInTerminal(buf)
+endfunc
+
+funct Test_diff_inline_multibuffer()
```
```suggestion
func Test_diff_inline_multibuffer()
```

ychin committed:
Code formatting cleanup per review

ychin committed:
Fix bug in diff_refine_inline_char_highlight assuming buffer 0 is valid

We need to make sure to use the first valid buffer, which may not be 0.
Just pass in the previously calculated buffer index to the algorithm.
Also update tests to check this as well.

chrisbra commented on runtime/doc/diff.txt:
```diff
@@ -226,14 +226,27 @@ The diffs are highlighted with these groups:
+|hl-DiffTextAdd|  DiffTextAdd	Added text inside a Changed line. Similar to
+				DiffText, but used when there is no
+				corresponding text in other buffers.  Will not
+				be used when `inline:` is set to "simple".
```
```suggestion
				be used when `inline:` is set to "simple" or "none".
```

ychin replied:
~~I don't think this is correct. If `inline:none` is set, *both* `DiffTextAdd` and `DiffText` are not going to be used. Only `DiffAdd` and `DiffChange` will be used.~~

~~Note that `DiffTextAdd` is a new highlight group and not the same as `DiffAdd`. This is intentional because in some colorschemes it would not work well if we just use `DiffAdd`  within a changed line.~~

Edit: nevermind. I misread the diff

ychin replied:
Done

chrisbra commented on runtime/doc/diff.txt:
```diff
@@ -226,14 +226,27 @@ The diffs are highlighted with these groups:
+				detailed diff between the changed blocks and
+				highlight the exact difference between the
+				two.  Will respect any 'diffopt' flag that
+				affects internal diff.
```
```suggestion
				affects internal diff.  Not used when `inline:` set to "none".
```

ychin replied:
Done

chrisbra commented on runtime/doc/options.txt:
```diff
@@ -2975,6 +2975,21 @@ A jump table for the options with a short description can be found at |Q_op|.
+				simple  Highlight from first different
+					character to the last one in each
+					line. This is the default if nothing
+					is set.
```
Hm, there doesn't seem to be a difference between, `:set diffopt=internal,inline:simple` and `:set diffopt=internal`. In other words, do we need the subvalue `simple`, since the same can be achieved when not setting `inline` at all?

ychin replied:
I feel like for an option with a list of possible values here, it's best for symmetry reasons to have all the values explicitly available. It's easier to parse in the docs and in auto-complete. It also makes it easier to change the defaults in the future without screwing people who explicitly set their `inline:` values. (E.g. we could change the defaults to `inline:char` under the hood in the future and it would "just work" if we have all the options all made available instead of relying on implicit settings)

This is also consistent with how `algorithm:` is implemented. `algorithm:myers` is the default, but we allow the user to explicitly set it, while it's also ok to not set the algorithm in `diffopt`, since this makes it easier on the user ergonomics to not having to specify an overly long verbose `diffopt`.

chrisbra replied:
Yeah, I came to the same conclusion when looking at the `algorithm` suboption. Makes sense. Thanks.

chrisbra commented on src/drawline.c:
```diff
@@ -1223,6 +1223,9 @@ win_line(
     int		change_start = MAXCOL;	// first col of changed area
     int		change_end = -1;	// last col of changed area
+    diffline_T	line_changes;
+    CLEAR_FIELD(line_changes);
```
That looks a bit strange between all the declarations, should this be moved to (above line 1481) instead?

ychin replied:
DONE

chrisbra commented on src/drawline.c:
```diff
@@ -2438,13 +2460,30 @@ win_line(
+			&& change_index < line_changes.num_changes - 1)
+		{
+		    if (ptr - line >= line_changes.changes[change_index+1].dc_start[line_changes.bufidx])
+		    {
```
we can remove the braces here

ychin replied:
Done

chrisbra commented on src/diff.c:
```diff
@@ -579,6 +591,12 @@ diff_alloc_new(tabpage_T *tp, diff_T *dprev, diff_T *dp)
 	dprev->df_next = dnew;
+
+    CLEAR_FIELD(dnew->df_lnum);
+    CLEAR_FIELD(dnew->df_count);
```
perhaps use `ALLOC_CLEAR_ONE(diff_T)` above and then get rid for `CLEAR_FIELD` here?

ychin replied:
Oh I didn't see that we could do that. Will do.

ychin replied:
Done.

chrisbra commented on src/diff.c:
```diff
@@ -3025,6 +3149,534 @@ diff_find_change(
+    // Use inline diff algorithm.
+    // The diff changes are usually cached so we check that first.
+    if (!dp->has_changes)
+    {
```
can leave out the braces

ychin replied:
Done

chrisbra commented on src/diff.c:
```diff
@@ -3025,6 +3149,534 @@ diff_find_change(
+	    dp = dp->df_next;
+    }
+    if (dp == NULL || diff_check_sanity(curtab, dp) == FAIL)
+    {
```
can leave out the braces

ychin replied:
Done

chrisbra commented on src/diff.c:
```diff
@@ -3025,6 +3149,534 @@ diff_find_change(
+    }
+
+    if (lnum - dp->df_lnum[idx] > INT_MAX)
+    {
```
same here

ychin replied:
Done

chrisbra commented on src/diff.c:
```diff
@@ -3025,6 +3149,534 @@ diff_find_change(
+	return FALSE;
+    }
+
+    if (lnum - dp->df_lnum[idx] > INT_MAX)
```
MAXLNUM?

ychin replied:
The reason for using `INT_MAX` is that `off` is an int, not `linenr_T`, in the original code for `diff_find_change()`. I guess I could have refactored `off` to use that instead, but just didn't feel like changing it. I could change it to `linenr_T` if you want.

chrisbra replied:
I think this is fine, thanks. Is this something we should document, that it is limited to INT_MAX changes per line?

ychin replied:
Hmm, do we really need to document this? That's a *lot* of lines... I feel like this is more a failsafe than a realistic user scenarios of a diff user and I mostly wrote the check to make sure we are safe against overflow bugs. I think this belongs more to implementation details category.

chrisbra replied:
Yeah, agree. It's an implementation detail. Let's keep it like this for now. Thanks!

ychin committed:
Misc code cleanups per review feedbacks

ychin committed:
Add 'inline:simple' to defaults, and allow multiple 'inline:' in diffopt

When multiple 'inline:' occurs in diffopt, just use the last one,
similar to 'algorithm:'.
